### PR TITLE
Add observer functions

### LINKED ISSUE #6491: Observable properties in gdscript
**Operating system or device - Godot version:**
gdscript feature

**Issue description** (what happened, and what was expected):
I find myself frequently writing accessors around properties in classes to notify outside observers when some property on a Node changes (e.g. The position of a Sprite, or any number of properties of a game entity that other consumers (like UI elements) would like to know).

It can become tedious and repetitive, but currently is worth it as you get the flexibility of information and changes flow/propagating through layers of the application naturally and in a decoupled fashion (interested consumers subscribe to data streams they're interested in), instead of being pushed by the data's owners (which can be fragile as you adapt application structure over time and the owner of some property has to keep track of who is interested in its contents over time).

It would be great if there were a pattern to abstract this away, akin to what you see with the Observable pattern in other languages/frameworks.

**Steps to reproduce:**
Currently:

```
# all of this for one observable, ouch!
signal sprite_pos_changed(new_pos)

var __sprite_pos = Vector2(0,0)
var sprite_pos = Vector2(0,0) setget sprite_pos_setter, sprite_pos_getter

func sprite_pos_setter(new_pos):
  __sprite_pos = new_pos
  emit_signal("sprite_pos_changed", new_pos)
func sprite_pos_getter():
  return __sprite_pos
```

This could be streamlined down to:

```
observable var sprite_pos = Vector2(0,0)
```

And, by convention, it would be understood that this would initialize a new user signal for the class with a specification of: `<property_name>_changed(new_val)`.

---

This is also possible, right now, by creating an Observable wrapper class to contain the value, but it makes things uglier when:

```
  var foo = self.bar
  self.bar = foo + 1
```

becomes:

```
  var foo = self.bar.get()
  self.bar.set(foo + 1)
```

Also, what's important about the Observable is that it's a data component of some larger data structure. Things like Observable wrappers just seek to atomize data in a way that can create liabilities down the road (like multiple data structures holding references to the same observable by accident, instead of a single "owner" with everyone else observing/subscripting to the data).

---

Additionally, with this pattern, you'll see it extended to collection types to fire an event when properties are added/removed to a collection (it should be noted that Array, Dictionary et al don't fire signals when their contents are modified and this is something you'd want to restrict to Nodes in the scene, anyways).

Also, once you have Observable collections, you can start modelling applications as handling streams of data over time, i.e. what's traditionally construed as "reactive programming". Just look at what's possible w/ RxPy (https://github.com/ReactiveX/RxPY)! I won't argue for/against-this as it's pretty advanced and something that requires a lot of discipline. There would be a lot of discovery to determine how it'd best fit into Godot (which is already very event driven, so these patterns would mostly be another approach to the same ideas/paradigms).

**Link to minimal example project** (optional but very welcome):
I know everyone is busy with lots of other great features and fixes for the community. This is not a high priority change, but I feel like it would be high value.

This is something that would greatly facilitate and make clear to members of the community how applications can be structured to allow data to propagating freely to interested consumers. The observer pattern allows data-owners to focus on their own logic, as opposed to worrying about implementation details of other entities interested in their contents.

Thanks for your time!

### PR DESCRIPTION
**Pull request by creikey**
- When there is a setter and an observer, value is not changed
   before calling the observer
 - Partially fixes #6491

### CODE REVIEW

**Comment by creikey** [comment]
Fixed formatting

**Comment by asheraryam** [comment]
In my humble opinion, this does not seem like a good idea.

First of all, it's standard practice for setters to have 1 and only 1 argument. 

Second of all, even if we keep in mind #6491, this does not seem like it will help a lot. If the setter takes no argument, then trying to do `onchange_value = 45 + 123` would **not** set the value and instead only do the extra you described in https://github.com/godotengine/godot/issues/6491#issuecomment-541495605, so it will only `print("updated")` but nothing will actually be updated

**Comment by creikey** [comment]
@asheraryam this PR will update the value independently of the setter function call, so the value is updated then the method is ran. This allows you to bind multiple variables to the same set method, as the method is only ran when they're set, not to set the variable.

File: modules/gdscript/gdscript.cpp [hunk 1]
```diff
@@ -949,7 +949,12 @@ bool GDScriptInstance::set(const StringName &p_name, const Variant &p_value) {
 			if (E->get().setter) {
 				const Variant *val = &p_value;
 				Variant::CallError err;
-				call(E->get().setter, &val, 1, err);
+				if (E->get().setter_no_set) {
+					call(E->get().setter, NULL, 0, err);
+					members.write[E->get().index] = p_value;
```
Comment (bojidar-bg) [inline modules/gdscript/gdscript.cpp]: Shouldn't you write to the variable before calling the "setter"?

**Comment by bojidar-bg** [comment]
I'm not sure about the idea -- on one hand, it allows reusing a function in many `setget`-s, which is a pretty cool thing; but on the other, it makes it possible to confuse a getter for a setter or vice-versa.

File: modules/gdscript/gdscript.cpp (hunk 1 above)
Comment (creikey) [reply modules/gdscript/gdscript.cpp]: I've written to the variable in the latest force push, thanks!

**Comment by creikey** [comment]
@bojidar-bg I'm a little uncertain as well, perhaps to make it more readable, a new `onchange` keyword could be added, so the syntax would look as such:

```gdscript
var foo = 0 setget set_foo , get_foo onchange foo_changed
var bar = 0 setget set_bar onchange bar_changed
var bee = 0 onchange bee_changed
```

**Comment by Calinou** [comment]
This could end up being confusing for people, so I'd be against it as well. An `onchange` keyword sounds better to me, to make this more explicit.

**Comment by groud** [comment]
> This could end up being confusing for people, so I'm against it as well. An onchange keyword sounds better to me, as it'd make the programmer's intent more explicit.

I completely agree, allowing setters with no arguments makes no sense.

**Comment by Calinou** [comment]
One downside I can see with an `onchange` keyword is that you can no longer have an associated getter, unlesss `onchange update_something setget ,get_some_property` is allowed.

**Comment by asheraryam** [comment]
>  this PR will update the value independently of the setter function call, so the value is updated then the method is ran.

The issue is that this is not always wanted, in many cases we want to perform some operation on the input before setting it, and removing that functionality would be against the standard implementation of setters that most languages have.

**Comment by bojidar-bg** [comment]
@asheraryam What this PR does applies only to zero-parameter setters, which were previously disallowed. It won't break setters with more than one parameter -- those would still get called, without the variable being set before or after them.

**Comment by creikey** [comment]
I'm currently in the process of updating this PR to use a new `observer` keyword to signify a function that is called when the property is changed as such:
```gdscript
var bar = 0 observer something_changed setget bar_set, bar_get
var fee = 0 setget ,fee_get observer something_changed
```
 In terms of the other half of the issue, maybe you can prepend the `var` with `broadcasting` to automatically make a signal for when that property is changed?

**Comment by creikey** [comment]
This pull request is now updated to include the `observer` syntax.

File: modules/gdscript/gdscript_tokenizer.cpp [hunk 2]
```diff
@@ -196,6 +197,7 @@ static const _kws _keyword_list[] = {
 	{ GDScriptTokenizer::TK_PR_TOOL, "tool" },
 	{ GDScriptTokenizer::TK_PR_STATIC, "static" },
 	{ GDScriptTokenizer::TK_PR_EXPORT, "export" },
+	{ GDScriptTokenizer::TK_PR_OBSERVER, "observer" },
```
Comment (bojidar-bg) [inline modules/gdscript/gdscript_tokenizer.cpp]: (non-blocking) `OBSERVER` is after `SETGET` in all other lists.

File: modules/gdscript/gdscript.cpp [hunk 3]
```diff
@@ -951,13 +951,28 @@ bool GDScriptInstance::set(const StringName &p_name, const Variant &p_value) {
 				Variant::CallError err;
 				call(E->get().setter, &val, 1, err);
 				if (err.error == Variant::CallError::CALL_OK) {
-					return true; //function exists, call was successful
+					if (E->get().observer) {
+						// call observer without changing data beforehand
```
Comment (bojidar-bg) [inline modules/gdscript/gdscript.cpp]: "Without"?

**Review by bojidar-bg — APPROVED** [review]
Code looks good overall

**Comment by bojidar-bg** [comment]
I guess the keyword could be called just `observe` instead of `observer`, not sure about the exact naming though.

**Comment by vnen** [comment]
Honestly I'd like this to through a proposal first so we can flesh out the usage and syntax. I dislike adding new keywords and the variable declaration syntax is already complex enough.

I know there has been discussion in the original issue, but I don't think we reached a consensus.

I also think this could be simplified with annotations (or "attributes" or "decorators", depending on what we decide to do at the end). It would avoid the stream of keywords to declare a variable.

**Comment by creikey** [comment]
@vnen I feel as if given the large amount of information ( both setter/getter methods and observer methods ), the syntax will either be too verbose for a scripting language or too unreadable, as the dense information will have been hidden with a "simplified" syntax ( to avoid the long list of keywords after a  variable ). 

Regardless, you're probably right in that there should be more discussion in the original issue, especially from more core contributors. What should be done with this PR until then?

**Comment by bojidar-bg** [comment]
I'd suggest leaving the PR open, but not changing it; the other option is to close it and mark it "salvageable". I guess you might use the feature in your projects, and when the consensus decision comes, update the PR and projects.

In any cases a proposal should be opened over at https://github.com/godotengine/godot-proposals/

**Comment by vnen** [comment]
This PR can be kept open for now.

What I really want to do is flesh out #18698 into a proper document, including the stuff we plan to do and what's nice to have. Of course, proposals are always welcome but we need a guideline of what we want with GDScript.

**Comment by creikey** [comment]
@vnen Yes, a GDScript language spec seems like it would be pretty useful. I feel like it would be best as a `.md` file that explicitly describes all the grammar and syntax in gdscript and what it does. I could start typing it up now and open a PR for it if that is desirable.

**Comment by aaronfranke** [comment]
@creikey Is this still desired? If so, it needs to be rebased on the latest master branch.

If not, we can close this and mark it as "salvageable".

**Comment by creikey** [comment]
@aaronfranke For 4.0 the parser is being rewritten by @vnen , and for future 3.2 releases @vnen stated that the proposal workflow is necessary to add a new keyword like this, so this work is not useful. Maybe mark as salvageable for those who want this feature in their custom fork?

**Comment by Calinou** [comment]
Closing, as this work needs to be discussed on the [Godot proposals](https://github.com/godotengine/godot-proposals) repository then redone against the new GDScript implementation in `master`.

**Pull request closed without merging**